optimizely/decision_service.py

```diff
@@ -670,7 +671,179 @@ def get_variation_for_feature(
+
+        Args:
+            holdout: The holdout configuration.
+            user_context: The user context.
+            project_config: The project config.
+
+        Returns:
+            A DecisionResult for the holdout.
+        """
+        from optimizely.helpers.enums import ExperimentAudienceEvaluationLogs
+
+        decide_reasons: list[str] = []
+        user_id = user_context.user_id
+        attributes = user_context.get_user_attributes()
+
+        if not holdout or not holdout.get('status') or holdout.get('status') != 'Running':
+            key = holdout.get('key') if holdout else 'unknown'
+            message = f"Holdout '{key}' is not running."
+            self.logger.info(message)
+            decide_reasons.append(message)
+            return {
+                'decision': Decision(None, None, None, None, None),
+                'error': False,
+                'reasons': decide_reasons
+            }
+
+        bucketing_id, bucketing_id_reasons = self._get_bucketing_id(user_id, attributes)
+        decide_reasons.extend(bucketing_id_reasons)
+
+        # Check audience conditions
+        audience_conditions = holdout.get('audienceIds')
+        user_meets_audience_conditions, reasons_received = audience_helper.does_user_meet_audience_conditions(
+            project_config,
+            audience_conditions,
+            ExperimentAudienceEvaluationLogs,
+            holdout.get('key', 'unknown'),
+            user_context,
+            self.logger
+        )
+        decide_reasons.extend(reasons_received)
+
+        if not user_meets_audience_conditions:
+            message = (
+                f"User '{user_id}' does not meet the conditions for holdout "
+                f"'{holdout['key']}'."
+            )
+            self.logger.debug(message)
+            decide_reasons.append(message)
+            return {
+                'decision': Decision(None, None, None, None, None),
+                'error': False,
+                'reasons': decide_reasons
+            }
+
+        # Bucket user into holdout variation
+        variation, bucket_reasons = self.bucketer.bucket(
+            project_config, holdout, user_id, bucketing_id  # type: ignore[arg-type]
+        )
+        decide_reasons.extend(bucket_reasons)
+
+        if variation:
+            # For holdouts, variation is a dict, not a Variation entity
+            variation_key = variation['key'] if isinstance(variation, dict) else variation.key
+            message = (
+                f"The user '{user_id}' is bucketed into variation '{variation_key}' "
+                f"of holdout '{holdout['key']}'."
+            )
+            self.logger.info(message)
+            decide_reasons.append(message)
+
+            # Create Decision with holdout - experiment is None, holdout field contains the holdout dict
+            holdout_decision: Decision = Decision(
+                experiment=None,
+                variation=None,
```

Comment:
I think variation here needs to have variation assigned.

holdout_decision: Decision = Decision(
      experiment=None,  # check if this should be None or holdout
      variation=variation,  **# Should contain the bucketed variation !!**
      source=enums.DecisionSources.HOLDOUT,
      cmab_uuid=None,
      holdout=holdout
  )

Javascript has this:

if (bucketResult.result) {
    const variation = configObj.variationIdMap[bucketResult.result];
    if (variation) {
      return {
        result: {
          experiment: holdout,  // Holdout as experiment
          variation: variation,  // Variation set
          decisionSource: DECISION_SOURCES.HOLDOUT
        },
        reasons: decideReasons
      };
    }
  }